### PR TITLE
Fix doc and test harnesses

### PR DESCRIPTION
**Pull request by basil-conto**
TL;DR:
* Byte-compiled `-third-item` is broken in several Emacs versions.
* Test harness currently fails in Emacs 27 due to unconditional loading of officially deprecated `cl.el`.
* Doc generation currently incorrectly replaces `'()` with `'nil`.
* Mentions of Marmalade were not properly removed in #302.

Please see commit messages for more details.

### CODE REVIEW

**Comment by basil-conto** [comment]
I will continue trying to fix the following error later:

`Error: the function `gv--defsetter' might not be defined at runtime.`

**Comment by basil-conto** [comment]
[BTW, my favourite solution would be to drop support for Emacs 24.1 and 24.2, if possible ;) ]

**Comment by basil-conto** [comment]
Phew, I think I got there in the end. In the process I discovered and fixed a bug in several Emacs versions:

1. `emacs -Q -batch -f batch-byte-compile dash.el`
2. `emacs -Q -l dash.elc`
3. ```el
   (-third-item ())
   ```
   <kbd>C-j</kbd>

This results in a `(void-function -third-item)` error because `dash.el` currently incorrectly assumes that loading `cl.el` at byte-compile-time will define `caddr` at runtime.

A future PR will improve usage of built-in and `cl-lib` `c[ad]+r` accessors, but this PR is just about getting byte-compilation, testing, and doc generation working consistently across the board.

File: dev/examples-to-docs.el [hunk 1]
```diff
@@ -153,8 +153,8 @@ FUNCTION may reference an elisp function, alias, macro or a subr."
 
 (defun simplify-quotes ()
   (goto-char (point-min))
-  (while (search-forward "(quote nil)" nil t)
-    (replace-match "'()"))
+  (while (re-search-forward (rx (| "'nil" "(quote nil)")) nil t)
```
Comment (Fuco1) [inline dev/examples-to-docs.el]: Tiny nitpick: I'd prefer `or` over `|`.  Also in the other files.

```suggestion
  (while (re-search-forward (rx (or "'nil" "(quote nil)")) nil t)
```

Comment (basil-conto) [reply dev/examples-to-docs.el]: Done.

**Comment by Fuco1** [comment]
Super!

**Pull request merged**